keyboards/kbd67/hotswap/keymaps/zunger/keymap.c

```diff
@@ -0,0 +1,178 @@
+   * |       | Α | Σ | Δ | Φ | Γ | Η | ¬ | Κ | Λ | ∪ | ∩ |       |    |
+   * |----------------------------------------------------------------|
+   * |        | Ζ | Ξ | Χ | Ψ | Β | Ν | Μ | ⊆ | ⊇ | ∉ |      | ⇑ | ⇔  |
+   * |----------------------------------------------------------------|
+   * |    |    |    |                       |   |    ||||| ⇐ | ⇓ | ⇒  |
+   * `----------------------------------------------------------------'
+  */
+[_SHIFTMAGIC] = LAYOUT(
+ KC_GRV,  H(2081), H(2082), H(2083), H(2084), H(2085), H(2086), H(2087), H(2088), H(2089), H(2080), H(208b), H(208a), H(2297), H(2248), _______, \
+ _______, H(2213), H(03a9), H(0395), H(03a1), H(03a4), H(0398), H(03a5), H(0399), H(039f), H(03a0), H(2204), H(221e), _______,          _______, \
+ _______, H(0391), H(03a3), H(0394), H(03a6), H(0393), H(0397), H(00ac), H(039a), H(039b), H(222a), H(2229),          _______,          _______, \
+ _______, H(0396), H(039e), H(03a7), H(03a8), H(0392), H(039d), H(039c), H(2286), H(2287), H(2209),                   _______, H(21d1), H(21d4), \
+ _______, _______, _______,                   _______,                            _______, _______,                   H(21d0), H(21d3), H(21d2)),
+};
+
+// Layer bitfields.
+#define MAGIC_LAYER (1UL << _MAGIC)
+#define SHIFTMAGIC_LAYER (1UL << _SHIFTMAGIC)
+// The layers we don't touch.
+#define LAYER_MASK ~(MAGIC_LAYER|SHIFTMAGIC_LAYER)
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  // Activate the appropriate magic layer:
+  //   User is holding AltGr+Shift => _SHIFTMAGIC
+  //   User is holding AltGr       => _MAGIC
+  //   Not holding AltGr           => None
+  // We do this the efficient way, by grabbing the non-magic bits from the current layer bitmap,
+  // setting the appropriate magic or shiftmagic bit, and doing an absolute layer set. (Layer
+  // sets are nontrivial, so we don't want to do extras!)
+
+  // We track shift and AltGr state on our own, because this function is called before get_mods is
+  // updated!
+  static bool shifted = false;
+  static bool magic = false;
+
+  if (keycode == KC_LSHIFT || keycode == KC_RSHIFT) {
+    shifted = record->event.pressed;
+  } else if (keycode == KC_RALT) {
+    magic = record->event.pressed;
+  }
+
+  // Update the layer.
+  const uint32_t old_layer_state = layer_state;
+  uint32_t new_layer_state = old_layer_state & LAYER_MASK;
+  if (magic) {
+    new_layer_state |= (shifted ? SHIFTMAGIC_LAYER : MAGIC_LAYER);
+  }
+  if (old_layer_state != new_layer_state) {
+    layer_state_set(new_layer_state);
+  }
+
+  // Second, implement SAFE_RESET. This is just a RESET key, but we want to avoid accidentally
+  // triggering it, so we also require left-shift. This means that to flash the keyboard you need
+  // to hit Fn+left-shift+esc, which is less likely to hit by accident than Fn+esc -- especially
+  // since AltGr+Esc is the easiest way to generate a backtick.
+  if (keycode == SAFE_RESET && record->event.pressed && shifted) {
```

Comment:
I don't know if you want to, but you don't actually need a new keycode for this.  You could just use `RESET`. 

The `process_record_user` function gets called before it's normally processed.  So, you could check for "shifted", and return false. 

Eg: 
```c
  if (keycode == RESET) {
    if (record->event.pressed && shifted) {
      return true; // to allow normal processing and actually reset it
    } else { 
      return false; // cancel additional action, we've handled it
    }
  }
```

https://github.com/qmk/qmk_firmware/blob/fdeec296369c0995bad1117e50946b2df84f35e4/quantum/quantum.c#L248

https://github.com/qmk/qmk_firmware/blob/fdeec296369c0995bad1117e50946b2df84f35e4/quantum/quantum.c#L292-L295

Reply:
Ah. That clearly works, but feels a bit scarily fragile to me -- it depends a lot on the order in which handlers run. Is that something stable enough that people should rely on it in code?

Reply:
Yes.  If you check the links above, `process_record_kb` and `process_record_kb` are always run before the core codes.  That's actually why I posted the links above (which show as code samples), so you could reference them yourself for independent verification. 

If this function returns false, then the core code returns false and doesn't run the built in code.

Reply:
K. If it's not a fragile invariant -- something likely to change unexpectedly in the future -- I'm all for simplicity. Done.

Reply:
It's a standard design pattern, where you trigger an event and allow the chained callbacks to cancel it instead of adding extra complexity to quell the event before it's raised. It's unlikely to change unless the entire system design does, so you can have some confidence in it.

Reply:
Yup. I just wanted to check that the order of the operations in `process_record_quantum` (e.g. that quantum keycodes are processed after process_record_user) is considered enough of an API guarantee that I don't have to worry about it being reshuffled without warning.